lib/Dialect/Handshake/HandshakeTypes.cpp

```diff
@@ -99,9 +103,153 @@ checkChannelExtra(function_ref<InFlightDiagnostic()> emitError,
   return success();
 }
 
+/// Parse the extra signals from the IR representation.
+/// Due to the restriction on the joint parser of ControlType and ChannelType,
+/// we don't parse square brackets here.
+/// For example: spec: i1, tag: i32
+/// This function is common to both ControlType and ChannelType.
+static LogicalResult
+parseExtraSignals(function_ref<InFlightDiagnostic()> emitError,
+                  AsmParser &odsParser,
+                  SmallVector<ExtraSignal::Storage> &extraSignalsStorage) {
+
+  auto parseSignal = [&]() -> ParseResult {
+    auto &signal = extraSignalsStorage.emplace_back();
+
+    if (odsParser.parseKeywordOrString(&signal.name) ||
+        odsParser.parseColon() || odsParser.parseType(signal.type))
+      return failure();
+
+    // Attempt to parse the optional upstream symbol
+    if (!odsParser.parseOptionalLParen()) {
+      std::string upstreamSymbol;
+      if (odsParser.parseKeywordOrString(&upstreamSymbol) ||
+          upstreamSymbol != UPSTREAM_SYMBOL || odsParser.parseRParen())
+        return failure();
+      // The default value of `signal.downstream` is true.
+      // I think this is a bit confusing.
+      // TODO: signal.upstream?
+      signal.downstream = false;
+    }
+    return success();
+  };
+
+  if (odsParser.parseCommaSeparatedList(parseSignal))
+    return failure();
+
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// ControlType
+//===----------------------------------------------------------------------===//
+
+void ControlType::print(AsmPrinter &odsPrinter) const {
+  odsPrinter << "<";
+  if (!getExtraSignals().empty()) {
+    printExtraSignals(odsPrinter, getExtraSignals());
+  }
+  odsPrinter << ">";
+}
+
+LogicalResult ControlType::verify(function_ref<InFlightDiagnostic()> emitError,
+                                  ArrayRef<ExtraSignal> extraSignals) {
+  return checkChannelExtra(emitError, extraSignals);
+}
+
+/// Parse the control type after "<[" is parsed.
+static Type parseControlAfterLSquare(AsmParser &odsParser) {
+  function_ref<InFlightDiagnostic()> emitError = [&]() {
+    return odsParser.emitError(odsParser.getCurrentLocation());
+  };
+
+  SmallVector<ExtraSignal::Storage> extraSignalsStorage;
+  if (failed(parseExtraSignals(emitError, odsParser, extraSignalsStorage)))
+    return {};
+
+  // Parse ']'
+  if (odsParser.parseRSquare())
+    return {};
+  // Parse literal '>'
+  if (odsParser.parseGreater())
+    return {};
+
+  // We convert the storage type to the non-storage type at the end of parsing.
+  // Be careful that the non-storage type just holds a raw string pointer inside
+  // (via StringRef), so passing it around may lead to dangling pointers.
+  SmallVector<ExtraSignal> extraSignals;
```

Comment:
I don't think these comments are helpful, these are general concepts more than relevant to this variable declaration.